### ISSUE TITLE
Fold Callers into CrossReferences [phab:D946]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1269]
_Created by **zarko** at 2016-06-29 23:09:54_: 

This diff adds callgraph support to (serving table-based)
cross-references. It is an intermediate step toward full
support as it still defers to the Slow* functions. The diff
also removes the old `callers` subcommand from the Kythe
command-line tool. Requesting callgraph information from the
`xrefs` subcommand is off by default.

The diff modifies CrossReferences API such that References
to callgraph-related nodes can be excluded or reported
exclusively. The former is useful because it reduces duplication
when asking for both references and callgraph data; the latter
is chiefly used internally.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

Just call it inside, as SlowCallers does below. It's less likely to be missed that way.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

Note that the convention of "begin with the name of the thing being documented" is mainly applicable to top-level functions and types that will appear in the godoc summary. It's not necessary to do it for locals. These could just be simple line-ending comments:

    // callsite anchor ticket to anchor details

    // semantic node ticket to owned anchor tickets

But also see line 771.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

"safe"?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

These aren't used before now—just make these `:=` and drop the declarations above.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

`if err := …; err != nil {`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

```
	defxrefs, err := service.CrossReferences(ctx, &xpb.CrossReferencesRequest{
		Ticket:         parentTickets,
		DefinitionKind: xpb.CrossReferencesRequest_ALL_DEFINITIONS,
		AnchorText:     true,
		PageSize:       math.MaxInt32,
	})
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

```
	relatedAnchors = append(relatedAnchors, &xpb.CrossReferencesReply_RelatedAnchor{
		Anchor:      callerAnchor,
		DisplayName: displayName,
		Site:        sites,
	})
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

`func(source, target, kind string) error {`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

Suggestion: `refSet` → `refs`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

Suggestion: `defxrefs` → `defs`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

Should this fail the request? The fact that there are more pages might deserve a warning, though.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

`// All callgraph-related references`

Please put a reference somewhere nearby to where it's documented what "callgraph-related" means.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

`// All references except those related to the callgraph.`

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:25:13_:

I find this name confusing. Suggestions: `NON_CALL_REFERENCES`, `ALL_EXCEPT_CALL`.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:55:12_:

Although consistent with its neighbors, use of output parameters like this is not idiomatic for Go. 

It's OK for now, since this is all unexported helper code, but we should figure out a better way to implement this.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 15:55:12_:

```
if !anchorText {
  a.Anchor.Text = ""
}
*to = append(*to, a)
```

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-11 20:15:46_:

  - Address comments.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 20:26:27_:

This check isn't needed; if there is no error, it will return a slice of the same length as the input.

(Though if you wanted to add that to the comment for same, it wouldn't hurt)

In fact, now that I look more closely, I think I'd just shortcut this entirely:

```
ticket, err := kytheuri.Fix(ticket)
if err != nil {
  return nil, err
}
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 20:26:27_:

I missed one.

```
	// This will not recursively call SlowCallersForCrossReferences as we're requesting NO_CALLERS above.
	xrefs, err := service.CrossReferences(ctx, &xpb.CrossReferencesRequest{
		Ticket:            xrefTickets,
		DefinitionKind:    xpb.CrossReferencesRequest_NO_DEFINITIONS,
		ReferenceKind:     xpb.CrossReferencesRequest_CALL_REFERENCES,
		DocumentationKind: xpb.CrossReferencesRequest_NO_DOCUMENTATION,
		CallerKind:        xpb.CrossReferencesRequest_NO_CALLERS,
		AnchorText:        true,
		PageSize:          math.MaxInt32,
	})
```

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2016-07-11 20:26:27_:

"noncall" (and at 211)

**Comment by kythe** [comment]
_**fromberger** wrote at 2016-07-11 20:26:27_:

Just a couple more small things, otherwise looks good to me.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-11 20:38:26_:

  - Address comments.
  - Address comments.

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2016-07-11 20:43:53_

**Comment by kythe** [comment]
_**zarko** closed via commit e7bb92f9b5769df09280142cd63308f3daa38756 at 2016-07-11 20:49:26_
